src/notification.cpp

```diff
@@ -52,6 +58,145 @@ NotificationManagerProxy *notificationManager()
+            stream << arg;
+            args.append(space + buffer.toBase64());
+            argsLength += args.last().length();
+        }
+
+        s.reserve(s.length() + argsLength);
+        foreach (const QString &arg, args) {
+            s.append(arg);
+        }
+    }
+
+    return s;
+}
+
+QStringList encodeActions(const QHash<QString, QString> &actions)
+{
+    QStringList rv;
+
+    // Actions are encoded as a sequence of name followed by displayName
+    QHash<QString, QString>::const_iterator it = actions.constBegin(), end = actions.constEnd();
+    for ( ; it != end; ++it) {
+        rv.append(it.key());
+        rv.append(it.value());
+    }
+
+    return rv;
+}
+
+QHash<QString, QString> decodeActions(const QStringList &actions)
+{
+    QHash<QString, QString> rv;
+
+    QStringList::const_iterator it = actions.constBegin(), end = actions.constEnd();
+    for ( ; it != end; ++it) {
+        const QString &key(*it);
+        if (++it == end) {
+            break;
+        }
+        rv.insert(key, *it);
+    }
+
+    return rv;
+}
+
+QPair<QHash<QString, QString>, QVariantHash> encodeActionHints(const QVariantList &actions)
+{
+    QPair<QHash<QString, QString>, QVariantHash> rv;
+
+    foreach (const QVariant &action, actions) {
+        QVariantMap vm = action.value<QVariantMap>();
+        const QString actionName = vm["name"].value<QString>();
+        if (!actionName.isEmpty()) {
+            const QString displayName = vm["displayName"].value<QString>();
+            const QString service = vm["service"].value<QString>();
+            const QString path = vm["path"].value<QString>();
+            const QString iface = vm["iface"].value<QString>();
+            const QString method = vm["method"].value<QString>();
+            const QVariantList arguments = vm["arguments"].value<QVariantList>();
+            const QString icon = vm["icon"].value<QString>();
+
+            rv.first.insert(actionName, displayName);
+            rv.second.insert(QString(HINT_REMOTE_ACTION_PREFIX) + actionName, encodeDBusCall(service, path, iface, method, arguments));
+            if (!icon.isEmpty()) {
+                rv.second.insert(QString(HINT_REMOTE_ACTION_ICON_PREFIX) + actionName, icon);
+            }
+        }
+    }
+
+    return rv;
+}
+
+QVariantList decodeActionHints(const QHash<QString, QString> &actions, const QVariantHash &hints)
+{
+    QVariantList rv;
+
+    QHash<QString, QString>::const_iterator ait = actions.constBegin(), aend = actions.constEnd();
+    for ( ; ait != aend; ++ait) {
+        const QString actionName = ait.key();
+        const QString displayName = ait.value();
+
+        const QString hintName = QString(HINT_REMOTE_ACTION_PREFIX) + actionName;
+        const QString &hint = hints[hintName].toString();
+        if (!hint.isEmpty()) {
+            QVariantMap action;
+
+            // Extract the element of the DBus call
+            QStringList elements(hint.split(' ', QString::SkipEmptyParts));
```

Comment:
If you `SkipEmptyParts` here, we should probably check for (and prevent) empty strings when assembling it above?